search/shards.go

```diff
@@ -816,15 +816,33 @@ func sendByRepository(result *zoekt.SearchResult, opts *zoekt.SearchOptions, sen
 
 	send := func(repoName string, a, b int, stats zoekt.Stats) {
 		index.SortFiles(result.Files[a:b])
+
+		// Filter RepoURLs and LineFragments to only those of repoName and its subRepositories
+		repoNames := []string{repoName}
+		for _, file := range result.Files[a:b] {
+			repoNames = append(repoNames, file.SubRepositoryName)
+		}
+
+		filteredRepoURLs := make(map[string]string)
+		filteredLineFragments := make(map[string]string)
+		for _, name := range repoNames {
+			if url, ok := result.RepoURLs[name]; ok {
+				filteredRepoURLs[name] = url
+			}
+			if frag, ok := result.LineFragments[name]; ok {
+				filteredLineFragments[name] = frag
+			}
+		}
```

Comment:
This code looks like it does quite a lot of allocating. Can you test this alternative implementation out.
```suggestion
		filteredRepoURLs := map[string]string{repoName: result.RepoURLs[repoName]}
		filteredLineFragments := map[string]string{repoName: result.LineFragments[repoName]}

		// Filter RepoURLs and LineFragments to only those of repoName and its
		// subRepositories if there are subRepositories
		for _, file := range result.Files[a:b] {
			name := file.SubRepositoryName
			if name == "" {
				continue
			}
			if _, ok := filteredRepoURLs[name]; ok {
				continue
			}
			if url, ok := result.RepoURLs[name]; ok {
				filteredRepoURLs[name] = url
			}
			if frag, ok := result.LineFragments[name]; ok {
				filteredLineFragments[name] = frag
			}
		}
```

Reply:
Oops... I intended to refactor `repoNames` into a set implementation after a quick test, totally slipped my mind, sorry for the oversight.

Your proposed code works well in my testing, I'd only suggest splitting the skip condition for potential future edge cases

```suggestion
		filteredRepoURLs := map[string]string{repoName: result.RepoURLs[repoName]}
		filteredLineFragments := map[string]string{repoName: result.LineFragments[repoName]}

		// Filter RepoURLs and LineFragments to only those of repoName and its
		// subRepositories if there are subRepositories
		for _, file := range result.Files[a:b] {
			name := file.SubRepositoryName
			if name == "" {
				continue
			}

			_, repoSet := filteredRepoURLs[name]
			url, ok := result.RepoURLs[name]
			if !repoSet && ok {
				filteredRepoURLs[name] = url
			}

			_, fragSet := filteredLineFragments[name]
			frag, ok := result.LineFragments[name]
			if !fragSet && ok {
				filteredLineFragments[name] = frag
			}
		}
```